ktor-client/ktor-client-features/ktor-client-auth/common/src/io/ktor/client/features/auth/providers/BearerAuthProvider.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.features.auth.providers
+
+import io.ktor.client.call.*
+import io.ktor.client.features.auth.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.http.auth.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.*
+
+/**
+ * Add [BearerAuthProvider] to client [Auth] providers.
+ */
+public fun Auth.bearer(block: BearerAuthConfig.() -> Unit) {
+    with(BearerAuthConfig().apply(block)) {
+        providers.add(BearerAuthProvider(_refreshTokens, _loadTokens, true, realm))
+    }
+}
+
+public class BearerTokens(
+    public val accessToken: String,
+    public val refreshToken: String
+)
+
+/**
+ * [BearerAuthProvider] configuration.
+ */
+public class BearerAuthConfig {
+    internal var _refreshTokens: suspend (call: HttpClientCall) -> BearerTokens? = { null }
+    internal var _loadTokens: suspend () -> BearerTokens? = { null }
+
+    public var realm: String? = null
+
+    public fun refreshTokens(block: suspend (call: HttpClientCall) -> BearerTokens?) {
```

Comment:
Would it be better to replace `HttpClientCall` with `HttpResponse`? If for some reason the user needs a call, he will be able to get it through `response.call` anyway. But most likely, the user is interested in the response and the `Authorization` header in it.